smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/PaginationGenerator.java

```diff
@@ -0,0 +1,199 @@
+        writer.addImport(inputSymbol.getName(),
+                inputSymbol.getName(),
+                inputSymbol.getNamespace());
+        writer.addImport(outputSymbol.getName(),
+                outputSymbol.getName(),
+                outputSymbol.getNamespace());
+        String nonModularLocation = serviceSymbol.getNamespace()
+                .replace(serviceSymbol.getName(), nonModularServiceName);
+        writer.addImport(nonModularServiceName,
+                nonModularServiceName,
+                nonModularLocation);
+        writer.addImport(serviceSymbol.getName(), serviceSymbol.getName(), serviceSymbol.getNamespace());
+
+        // Import Pagination types
+        writer.addImport("Paginator", "Paginator", "@aws-sdk/types");
+        writer.addImport(paginationType, paginationType, "./" + PAGINATION_INTERFACE_FILE.replace(".ts", ""));
+
+        this.writeClientSideRequest();
+        this.writeRequest();
+        this.writePager();
+    }
+
+    static String getOutputFilelocation(OperationShape operation) {
+        return "pagination/" + operation.getId().getName() + "Paginator.ts";
+    }
+
+    static void generateServicePaginationInterfaces(String nonModularServiceName,
+                                                           Symbol service,
+                                                           TypeScriptWriter writer) {
+        writer.addImport("PaginationConfiguration", "PaginationConfiguration", "@aws-sdk/types");
+        String nonModularLocation = service.getNamespace().replace(service.getName(), nonModularServiceName);
+        writer.addImport(nonModularServiceName, nonModularServiceName, nonModularLocation);
+        writer.addImport(service.getName(), service.getName(), service.getNamespace());
+
+        writer.openBlock("export interface $LPaginationConfiguration extends PaginationConfiguration {",
+                "}", nonModularServiceName, () -> {
+            writer.write("client: $L | $L;", nonModularServiceName, service.getName());
+        });
+    }
+
+    private void writePager() {
+        String serviceTypeName = serviceSymbol.getName();
+        String inputTypeName = inputSymbol.getName();
+        String outputTypeName = outputSymbol.getName();
+        String inputTokenName = paginatedInfo.getInputTokenMember().getMemberName();
+        String outputTokenName = paginatedInfo.getOutputTokenMember().getMemberName();
+
+        writer.openBlock(
+                "export async function* $LPaginate(config: $L, input: $L, ...additionalArguments: any): Paginator<$L>{",
+                "}",  methodName, paginationType, inputTypeName, outputTypeName, () -> {
+            writer.write("let token: string | undefined = config.startingToken || '';");
+
+            writer.write("let hasNext = true;");
+            writer.write("let page:$L;", outputTypeName);
+            writer.openBlock("while (hasNext) {", "}", () -> {
+                writer.write("input[\"$L\"] = token;", inputTokenName);
```

Comment:
Cleanup: all the instances of `\"$L\"` can be replaced by `$S` to get a string representation.